src/sage/matrix/matrix2.pyx

```diff
@@ -18656,8 +18660,14 @@ def _matrix_power_symbolic(A, n):
         # D^i(f) / i! with f = x^n and D = differentiation wrt x
         if hasattr(mk, 'radical_expression'):
             mk = mk.radical_expression()
-        vk = [(binomial(n, i) * mk**(n-i)).simplify_full()
-              for i in range(nk)]
+
+        # Return mk^(n-i) instead of mk**(n-i) if mk=0
+        if mk:
+            vk = [(binomial(n, i) * mk._pow_(n-i)).simplify_full()
+                  for i in range(nk)]
+        else:
```

Comment:
In fact, I realize that after the computation, we can get a value even for negative n (or even also for non-integer n).

I think that, somehow, the system should raise an error if mk=0 but the evaluation is made at bad values of n. I have no idea on how to do that properly.

Reply:
This can be done by declaring variable with domain="integer", here when you do not give any domain to a variable it will consider it in complex plane by default. 

But we can check whether the variable have an integer value of not at the time of evaluation.

> I think that, somehow, the system should raise an error if mk=0 but the evaluation is made at bad values of n.

I didn't get that, why system should give an error if mk=0 ?

Reply:
I'll implement this in a new PR as this topic is not concentrated on nilpotent matrices.

Reply:
> This can be done by declaring variable with domain="integer", here when you do not give any domain to a variable it will consider it in complex plane by default.
> 

great!

> But we can check whether the variable have an integer value of not at the time of evaluation.
> 
> > I think that, somehow, the system should raise an error if mk=0 but the evaluation is made at bad values of n.
> 
> I didn't get that, why system should give an error if mk=0 ?

It should raise an error if mk=0 and n negative since the matrix is then non-invertible. For n positive non-integer, I don't know if it makes sense either.

Reply:
> I'll implement this in a new PR as this topic is not concentrated on nilpotent matrices.

By the way, I already opened an other issue #36863 for general simplification of `0^n`. 

But the problem for matrices is somehow different since: 

- the exponent is generally assumed to be an integer
- the general formula for the power of a Jordan block with eigenvalue 0 involves some expression of the form `P(n)*0^{n-i}` (P polynomial)  where the meaning of `0^{n-i}`for `0<i <=n`should be understood as `kronecker_delta(i,n)` in this special case (assuming that n is a positive integer)

Reply:
> In fact, I realize that after the computation, we can get a value even for negative n (or even also for non-integer n).
> I think that, somehow, the system should raise an error if mk=0 but the evaluation is made at bad values of n. I have no idea on how to do that properly.

I checked what I can do to determine whether we are evaluating an integer power or not. I found that when evaluating the matrix, we only replace the variables with their values, and we cannot impose a condition that the variable should be an integer for obvious reasons because we are not sure that the variable is the symbolic power only. The only place where we can make a change is when calculating the power of the matrix (i.e., A^n). However, if we do so, it will not be ideal because we can define a real power of a matrix, and imposing a condition would not be appropriate. So, I think implementing a system for this is not practical. So it is better to use domain=integer while defining a variable.